internal/pkg/service/stream/storage/repository/slice/slice_state.go

```diff
@@ -0,0 +1,149 @@
+	var file model.File
+	var old, updated model.Slice
+	return op.Atomic(r.client, &updated).
+		// Read file
+		Read(func(ctx context.Context) op.Op {
+			return r.files.Get(k.FileKey).WithResultTo(&file)
+		}).
+		// Read entity for modification
+		Read(func(ctx context.Context) op.Op {
+			return r.Get(k).WithResultTo(&old)
+		}).
+		// Update the entity
+		Write(func(ctx context.Context) op.Op {
+			return r.switchState(ctx, file.State, old, now, from, to).SetResultTo(&updated)
+		})
+}
+
+func (r *Repository) switchState(ctx context.Context, fileState model.FileState, oldValue model.Slice, now time.Time, from, to model.SliceState) *op.TxnOp[model.Slice] {
+	// Validate from state
+	if oldValue.State != from {
+		return op.ErrorTxn[model.Slice](errors.Errorf(`slice "%s" is in "%s" state, expected "%s"`, oldValue.SliceKey, oldValue.State, from))
+	}
+
+	// Validate file and slice state combination
+	if err := validateFileAndSliceState(fileState, to); err != nil {
+		return op.ErrorTxn[model.Slice](errors.PrefixErrorf(err, `unexpected slice "%s" state:`, oldValue.SliceKey))
+	}
+
+	// Switch slice state
+	newValue, err := oldValue.WithState(now, to)
+	if err != nil {
+		return op.ErrorTxn[model.Slice](err)
+	}
+
+	return r.save(ctx, now, &oldValue, &newValue)
+}
+
+func (r *Repository) switchStateInBatch(ctx context.Context, fState model.FileState, original []model.Slice, now time.Time, from, to model.SliceState) *op.TxnOp[[]model.Slice] {
+	var updated []model.Slice
+	txn := op.TxnWithResult(r.client, &updated)
+	for _, slice := range original {
+		txn.Merge(r.switchState(ctx, fState, slice, now, from, to).OnSucceeded(func(r *op.TxnResult[model.Slice]) {
+			updated = append(updated, r.Result())
+		}))
+	}
+	return txn
+}
+
+func (r *Repository) switchSlicesToImported(file model.File, now time.Time) *op.AtomicOp[[]model.Slice] {
+	var slices, updated []model.Slice
+	return op.Atomic(r.client, &updated).
+		// Load slices
+		Read(func(ctx context.Context) op.Op {
+			return r.ListIn(file.FileKey).WithAllTo(&slices)
+		}).
+		// Close slices
+		Write(func(ctx context.Context) op.Op {
+			return r.switchStateInBatch(ctx, file.State, slices, now, model.SliceUploaded, model.SliceImported).SetResultTo(&updated)
+		})
+}
+
+func (r *Repository) validateSliceStates(file model.File) *op.AtomicOp[[]model.Slice] {
+	var slices, updated []model.Slice
+	return op.Atomic(r.client, &updated).
+		// Load slices
+		Read(func(ctx context.Context) op.Op {
+			return r.ListIn(file.FileKey).WithAllTo(&slices)
+		}).
+		// Validate slices states
+		Write(func(ctx context.Context) op.Op {
+			for _, slice := range slices {
+				if err := validateFileAndSliceState(file.State, slice.State); err != nil {
+					return op.ErrorOp(err)
+				}
+			}
+			return nil
+		})
+}
+
+// ValidateFileAndSliceState validates combination of file and slice state.
+func validateFileAndSliceState(fileState model.FileState, sliceState model.SliceState) error {
```

Comment:
Could we do documentation of various transitions for slices and file  states? E.g file `writing` -> `closing` is valid, not reversible etc. For slice I see that `uploaded` -> `imported` is valid state transition.

Reply:
File states: https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-596-file-repo/internal/pkg/service/stream/storage/model/filestate.go#L12-L48

Slice states: https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-596-file-repo/internal/pkg/service/stream/storage/model/slicestate.go

I added there one more comment:
https://github.com/keboola/keboola-as-code/pull/1791/commits/9c1b083f141e243d769899f948198a36559d30e4